StitchCore/StitchCoreTests/Admin/StitchAdminClient.swift

```diff
@@ -0,0 +1,186 @@
+import Foundation
+import PromiseKit
+import ExtendedJson
+@testable import StitchCore
+
+// Any endpoint that can be described with basic
+// CRUD operations
+internal protocol Endpoint {
+    /// absolute url to this endpoint
+    var url: String { get }
+    // stitch http client for making requests
+    var httpClient: StitchHTTPClient { get }
+}
+
+// Adds an endpoint method that GETs some list
+internal protocol List: Endpoint { associatedtype Model: Decodable }
+extension List {
+    public func list() -> Promise<[Model]> {
+        return self.httpClient.doRequest {
+            $0.endpoint = self.url
+        }.flatMap {
+            return try JSONDecoder().decode([Model].self,
+                                            from: JSONSerialization.data(withJSONObject: $0))
+        }
+    }
+}
+
+// Adds an endpoint method that GETs some id
+internal protocol Get: Endpoint { associatedtype Model: Decodable }
+extension Get {
+    public func get() -> Promise<Model> {
+        return self.httpClient.doRequest {
+            $0.endpoint = self.url
+        }.flatMap {
+            return try JSONDecoder().decode(Model.self,
+                                            from: JSONSerialization.data(withJSONObject: $0))
+        }
+    }
+}
+
+// Adds an endpoint method that DELETEs some id
+internal protocol Remove: Endpoint { }
+extension Remove {
+    public func remove() -> Promise<Any> {
+        return self.httpClient.doRequest {
+            $0.endpoint = self.url
+            $0.method = .delete
+        }
+    }
+}
+
+// Adds an endpoint method that POSTs new data
```

Comment:
[harsh-opinion] I'm not convinced that HTTP methods and actions should map to protocols. They are saving code, but that could still easily be done with a more succinct doRequest interface. I feel like this adds needless complexity and is slightly an abuse of protocols despite it looking cool.

This opinion applies to `Endpoint`, `List`, `Get`, `Remove`, `Create`, `Update`, `Enable`, and `Disable`